go/types/list_editor.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2017 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package types
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/attic-labs/noms/go/d"
+)
+
+type ListEditor struct {
+	l     List
+	edits *listEdit
+}
+
+func NewListEditor(l List) *ListEditor {
+	return &ListEditor{l, nil}
+}
+
+func (le *ListEditor) Kind() NomsKind {
+	return ListKind
+}
+
+func (le *ListEditor) Value(vrw ValueReadWriter) Value {
+	return le.List(vrw)
+}
+
+func (le *ListEditor) List(vrw ValueReadWriter) List {
+	if le.edits == nil {
+		return le.l // no edits
+	}
+
+	seq := le.l.sequence()
+	vr := seq.valueReader()
+
+	if vrw != nil {
+		vr = vrw
+	}
+
+	cursChan := make(chan chan *sequenceCursor)
+	spliceChan := make(chan chan listEdit)
+
+	go func() {
+		for edit := le.edits; edit != nil; edit = edit.next {
+			edit := edit
+
+			// TODO: Use ReadMany
+			cc := make(chan *sequenceCursor, 1)
+			cursChan <- cc
+
+			go func() {
+				cc <- newCursorAtIndex(seq, edit.idx, false)
+			}()
+
+			sc := make(chan listEdit, 1)
+			spliceChan <- sc
+
+			wg := sync.WaitGroup{}
+			subEditors := false
+			for i, v := range edit.inserted {
+				if _, ok := v.(Value); ok {
+					continue
+				}
+
+				subEditors = true
+				idx := i
+				wg.Add(1)
+				go func() {
+					edit.inserted[idx] = v.Value(vrw)
+					wg.Done()
+				}()
+			}
+			if !subEditors {
+				sc <- *edit
+				continue
+			}
+
+			go func() {
+				wg.Wait()
+				sc <- *edit
+			}()
+		}
+
+		close(cursChan)
+		close(spliceChan)
+	}()
+
+	var ch *sequenceChunker
+	for cc := range cursChan {
+		cur := <-cc
+		sp := <-<-spliceChan
+
+		if ch == nil {
+			ch = newSequenceChunker(cur, 0, vr, vrw, makeListLeafChunkFn(vr), newIndexedMetaSequenceChunkFn(ListKind, vr), hashValueBytes)
+		} else {
+			ch.advanceTo(cur)
+		}
+
+		dc := sp.removed
+		for dc > 0 {
+			ch.Skip()
+			dc--
+		}
+
+		for _, v := range sp.inserted {
+			if emp, ok := v.(Emptyable); ok && emp.Empty() {
```

Comment:
This comes up when you are mutating graphs and you discover that some collection *has become empty*. All of examples we had, removed the empty container in that case. Very strange bugs could arise if you didn't do that (at least in those cases).